From 45fe3869aaf8bc74bdca5fd96bd9c6ff85a19a83 Mon Sep 17 00:00:00 2001
From: Artur Puzio <contact@puzio.waw.pl>
Date: Tue, 28 Jul 2020 23:25:29 +0200
Subject: [PATCH] Fix removing from PCIBus devices in unplug_disks

Missing remove from devices table in PCIBus causes null pointer
dereference when qemu iterates over that table in some cases
---
 hw/i386/xen/xen_platform.c | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/hw/i386/xen/xen_platform.c b/hw/i386/xen/xen_platform.c
index 0f7b05e5e1..c518631e24 100644
--- a/hw/i386/xen/xen_platform.c
+++ b/hw/i386/xen/xen_platform.c
@@ -27,6 +27,7 @@
 #include "qapi/error.h"
 #include "hw/ide.h"
 #include "hw/pci/pci.h"
+#include "hw/pci/pci_bus.h"
 #include "hw/irq.h"
 #include "hw/xen/xen_common.h"
 #include "migration/vmstate.h"
@@ -151,17 +152,20 @@ static void unplug_disks(PCIBus *b, PCIDevice *d, void *opaque)
 
     switch (pci_get_word(d->config + PCI_CLASS_DEVICE)) {
     case PCI_CLASS_STORAGE_IDE:
+        b->devices[d->devfn] = NULL;
         pci_piix3_xen_ide_unplug(DEVICE(d), aux);
         break;
 
     case PCI_CLASS_STORAGE_SCSI:
         if (!aux) {
+            b->devices[d->devfn] = NULL;
             object_unparent(OBJECT(d));
         }
         break;
 
     case PCI_CLASS_STORAGE_EXPRESS:
         if (flags & UNPLUG_NVME_DISKS) {
+            b->devices[d->devfn] = NULL;
             object_unparent(OBJECT(d));
         }
 
-- 
2.27.0

